libs/@local/hashql/core/src/arena/transaction.rs

```diff
@@ -0,0 +1,510 @@
+use core::ops::Index;
+
+use crate::id::{HasId, Id as _};
+/// A saved state of a `TransactionalArena` that can be restored at a later time.
+///
+/// Checkpoints allow for transactional operations by saving the state of the arena
+/// before making changes, and then restoring it if those changes need to be rolled back.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Checkpoint<T>(rpds::Vector<T>);
+
+/// A persistent arena that provides transactional semantics for storing and modifying objects.
+///
+/// `TransactionalArena` uses a persistent data structure internally which allows efficient
+/// creation of checkpoints and restoration to previous states. This makes it suitable for
+/// scenarios where you need to try a series of operations and potentially roll them back.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct TransactionalArena<T> {
+    items: rpds::Vector<T>,
+}
+
+impl<T> TransactionalArena<T> {
+    /// Creates a new, empty `TransactionalArena`.
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            items: rpds::Vector::new(),
+        }
+    }
+
+    /// Returns the next available ID for an item in this arena.
+    fn next_id(&self) -> usize {
+        self.items.len()
+    }
+
+    /// Adds an item to the arena using a builder function that receives the assigned ID.
+    ///
+    /// This is useful when the item's constructor needs to know its own ID.
+    ///
+    /// # Returns
+    ///
+    /// The ID assigned to the newly added item.
+    pub fn push_with(&mut self, item: impl FnOnce(T::Id) -> T) -> T::Id
+    where
+        T: HasId,
+    {
+        let id = T::Id::from_usize(self.next_id());
+
+        self.items.push_back_mut(item(id));
+        id
+    }
+
+    /// Adds an existing item to the arena.
+    ///
+    /// # Returns
+    ///
+    /// The ID assigned to the newly added item.
+    pub fn push(&mut self, item: T) -> T::Id
+    where
+        T: HasId,
+    {
+        let id = T::Id::from_usize(self.next_id());
+
+        self.items.push_back_mut(item);
+        id
+    }
+
+    /// Updates an item in the arena.
+    ///
+    /// # Panics
+    ///
+    /// If the item's ID is not found in the arena.
+    pub fn update(&mut self, item: T)
+    where
+        T: HasId,
+    {
+        let id = item.id();
+
+        let inserted = self.items.set_mut(id.as_usize(), item);
+        assert!(inserted, "Item with id {id} not found");
+    }
+
+    /// Updates an item in the arena using the provided closure.
+    ///
+    /// # Panics
+    ///
+    /// If the item with the given ID is not found in the arena.
+    pub fn update_with(&mut self, id: T::Id, closure: impl FnOnce(&mut T))
+    where
+        T: HasId + Clone,
```

Comment:
Why we need `Clone`?

Reply:
explained in https://github.com/hashintel/hash/pull/6918#discussion_r2043837651